File: Packages/OpenVibbleKit/Sources/BuddyProtocol/BridgeModels.swift
import Foundation

public struct BridgeAck: Codable, Equatable, Sendable {
    public let ack: String
    public let ok: Bool
    public let n: Int?
    public let error: String?
    public let data: JSONValue?

    public init(ack: String, ok: Bool, n: Int? = nil, error: String? = nil, data: JSONValue? = nil) {
        self.ack = ack
        self.ok = ok
        self.n = n
        self.error = error
        self.data = data
    }
}

public struct TimeSync: Codable, Equatable, Sendable {
    public let epochSeconds: Int64
    public let timezoneOffsetSeconds: Int

    public init(epochSeconds: Int64, timezoneOffsetSeconds: Int) {
        self.epochSeconds = epochSeconds
        self.timezoneOffsetSeconds = timezoneOffsetSeconds
    }

    enum CodingKeys: String, CodingKey {
        case time
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        var values = try container.nestedUnkeyedContainer(forKey: .time)
        let epoch = try values.decode(Int64.self)
        let timezone = try values.decode(Int.self)
        self.init(epochSeconds: epoch, timezoneOffsetSeconds: timezone)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        var values = container.nestedUnkeyedContainer(forKey: .time)
        try values.encode(epochSeconds)
        try values.encode(timezoneOffsetSeconds)
    }
}

public struct HeartbeatPrompt: Codable, Equatable, Sendable {
    public let id: String
    public let tool: String?
    public let hint: String?

    public init(id: String, tool: String?, hint: String?) {
        self.id = id
        self.tool = tool
        self.hint = hint
    }
}

public struct HeartbeatSnapshot: Codable, Equatable, Sendable {
    public let total: Int
    public let running: Int
    public let waiting: Int
    public let msg: String
    public let entries: [String]
    public let tokens: Int?
    public let tokensToday: Int?
    public let prompt: HeartbeatPrompt?
    public let completed: Bool?

    enum CodingKeys: String, CodingKey {
        case total, running, waiting, msg, entries, tokens, prompt, completed
        case tokensToday = "tokens_today"
    }

    public init(
        total: Int,
        running: Int,
        waiting: Int,
        msg: String,
        entries: [String],
        tokens: Int?,
        tokensToday: Int?,
        prompt: HeartbeatPrompt?,
        completed: Bool? = nil
    ) {
        self.total = total
        self.running = running
        self.waiting = waiting
        self.msg = msg
        self.entries = entries
        self.tokens = tokens
        self.tokensToday = tokensToday
        self.prompt = prompt
        self.completed = completed
    }
}

public struct TurnEvent: Codable, Equatable, Sendable {
    public let evt: String
    public let role: String
    public let content: [JSONValue]

    public init(evt: String, role: String, content: [JSONValue]) {
        self.evt = evt
        self.role = role
        self.content = content
    }
}

public enum PermissionDecision: String, Codable, Equatable, Sendable {
    case once
    case deny
}

public struct PermissionCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let id: String
    public let decision: PermissionDecision

    public init(id: String, decision: PermissionDecision) {
        self.cmd = "permission"
        self.id = id
        self.decision = decision
    }
}

public struct NameCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let name: String

    public init(name: String) {
        self.cmd = "name"
        self.name = name
    }
}

public struct OwnerCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let name: String

    public init(name: String) {
        self.cmd = "owner"
        self.name = name
    }
}

public struct StatusCommand: Codable, Equatable, Sendable {
    public let cmd: String

    public init() {
        self.cmd = "status"
    }
}

public struct UnpairCommand: Codable, Equatable, Sendable {
    public let cmd: String

    public init() {
        self.cmd = "unpair"
    }
}

public struct CharBeginCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let name: String
    public let total: Int

    public init(name: String, total: Int) {
        self.cmd = "char_begin"
        self.name = name
        self.total = total
    }
}

public struct FileCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let path: String
    public let size: Int

    public init(path: String, size: Int) {
        self.cmd = "file"
        self.path = path
        self.size = size
    }
}

public struct ChunkCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let d: String

    public init(base64: String) {
        self.cmd = "chunk"
        self.d = base64
    }
}

public struct FileEndCommand: Codable, Equatable, Sendable {
    public let cmd: String

    public init() {
        self.cmd = "file_end"
    }
}

public struct CharEndCommand: Codable, Equatable, Sendable {
    public let cmd: String

    public init() {
        self.cmd = "char_end"
    }
}

public struct SpeciesCommand: Codable, Equatable, Sendable {
    public let cmd: String
    public let idx: Int

    public init(idx: Int) {
        self.cmd = "species"
        self.idx = idx
    }
}

public enum BridgeCommand: Equatable, Sendable {
    case status
    case name(String)
    case owner(String)
    case unpair
    case charBegin(name: String, total: Int)
    case file(path: String, size: Int)
    case chunk(base64: String)
    case fileEnd
    case charEnd
    case permission(id: String, decision: PermissionDecision)
    case species(idx: Int)
    case unknown(String)
}

public enum BridgeInboundMessage: Equatable, Sendable {
    case heartbeat(HeartbeatSnapshot)
    case turn(TurnEvent)
    case timeSync(TimeSync)
    case command(BridgeCommand)
}

public enum BridgeProtocolError: Error, Equatable {
    case invalidEnvelope
}

File: android/buddy-protocol/src/main/kotlin/com/openvibble/protocol/BridgeModels.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.openvibble.protocol

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonElement

@Serializable
data class BridgeAck(
    val ack: String,
    val ok: Boolean,
    val n: Int? = null,
    val error: String? = null,
    val data: JsonElement? = null,
)

/** Envelope `{"time":[epochSeconds, timezoneOffsetSeconds]}`. */
@Serializable(with = TimeSync.Companion::class)
data class TimeSync(
    val epochSeconds: Long,
    val timezoneOffsetSeconds: Int,
) {
    companion object : KSerializer<TimeSync> {
        override val descriptor: SerialDescriptor =
            buildClassSerialDescriptor("TimeSync") {
                element<List<Long>>("time")
            }

        override fun deserialize(decoder: Decoder): TimeSync {
            val element = decoder.decodeJsonElementCompat()
            val arr = element.asJsonObject()["time"]?.asJsonArray()
                ?: error("TimeSync missing 'time' array")
            val epoch = arr[0].asJsonPrimitive().content.toLong()
            val tz = arr[1].asJsonPrimitive().content.toInt()
            return TimeSync(epoch, tz)
        }

        override fun serialize(encoder: Encoder, value: TimeSync) {
            val obj = kotlinx.serialization.json.buildJsonObject {
                put(
                    "time",
                    kotlinx.serialization.json.buildJsonArray {
                        add(kotlinx.serialization.json.JsonPrimitive(value.epochSeconds))
                        add(kotlinx.serialization.json.JsonPrimitive(value.timezoneOffsetSeconds))
                    },
                )
            }
            encoder.encodeJsonElementCompat(obj)
        }
    }
}

@Serializable
data class HeartbeatPrompt(
    val id: String,
    val tool: String? = null,
    val hint: String? = null,
)

@Serializable
data class HeartbeatSnapshot(
    val total: Int,
    val running: Int,
    val waiting: Int,
    val msg: String,
    val entries: List<String> = emptyList(),
    val tokens: Int? = null,
    @SerialName("tokens_today") val tokensToday: Int? = null,
    val prompt: HeartbeatPrompt? = null,
    val completed: Boolean? = null,
)

@Serializable
data class TurnEvent(
    val evt: String,
    val role: String,
    val content: List<JsonElement> = emptyList(),
)

@Serializable
enum class PermissionDecision {
    @SerialName("once") ONCE,
    @SerialName("deny") DENY,
}

@Serializable
data class PermissionCommand(
    val cmd: String = "permission",
    val id: String,
    val decision: PermissionDecision,
)

@Serializable
data class NameCommand(val cmd: String = "name", val name: String)

@Serializable
data class OwnerCommand(val cmd: String = "owner", val name: String)

@Serializable
data class StatusCommand(val cmd: String = "status")

@Serializable
data class UnpairCommand(val cmd: String = "unpair")

@Serializable
data class CharBeginCommand(val cmd: String = "char_begin", val name: String, val total: Int)

@Serializable
data class FileCommand(val cmd: String = "file", val path: String, val size: Int)

@Serializable
data class ChunkCommand(val cmd: String = "chunk", val d: String) {
    companion object {
        fun fromBase64(base64: String): ChunkCommand = ChunkCommand(d = base64)
    }
}

@Serializable
data class FileEndCommand(val cmd: String = "file_end")

@Serializable
data class CharEndCommand(val cmd: String = "char_end")

@Serializable
data class SpeciesCommand(val cmd: String = "species", val idx: Int)

sealed class BridgeCommand {
    data object Status : BridgeCommand()
    data class Name(val name: String) : BridgeCommand()
    data class Owner(val name: String) : BridgeCommand()
    data object Unpair : BridgeCommand()
    data class CharBegin(val name: String, val total: Int) : BridgeCommand()
    data class File(val path: String, val size: Int) : BridgeCommand()
    data class Chunk(val base64: String) : BridgeCommand()
    data object FileEnd : BridgeCommand()
    data object CharEnd : BridgeCommand()
    data class Permission(val id: String, val decision: PermissionDecision) : BridgeCommand()
    data class Species(val idx: Int) : BridgeCommand()
    data class Unknown(val cmd: String) : BridgeCommand()
}

sealed class BridgeInboundMessage {
    data class Heartbeat(val snapshot: HeartbeatSnapshot) : BridgeInboundMessage()
    data class Turn(val event: TurnEvent) : BridgeInboundMessage()
    data class Time(val sync: TimeSync) : BridgeInboundMessage()
    data class Command(val command: BridgeCommand) : BridgeInboundMessage()
}

class BridgeProtocolException(message: String) : RuntimeException(message)
